packages/access-api/src/models/delegations.js

```diff
@@ -179,3 +195,160 @@ export function delegationsTableBytesToArrayBuffer(sqlValue) {
     return Uint8Array.from(sqlValue)
   }
 }
+
+/**
+ * @typedef {`delegations_v3`} DelegationsTableWithoutBytesName
+ */
+
+/** @type {DelegationsTableWithoutBytesName} */
+export const delegationsV3Table = `delegations_v3`
+
+/**
+ * @typedef {import('../types/access-api-cf-db').AccessApiD1TablesV2} AccessApiD1TablesV2
+ * @typedef {import('../types/access-api-cf-db').AccessApiD1TablesV3} AccessApiD1TablesV3
+ */
+
+export class DbDelegationsStorageWithR2 {
+  // @todo abstract away R2 specifics into DagStore: ~AsyncMap<CID, Ucanto.Delegation>
+  /** @type {R2Bucket} */
+  #dags
+  /** @type {import('../types/database').Database<AccessApiD1TablesV3>} */
+  #db
+  /** @type {keyof AccessApiD1TablesV3} */
+  #delegationsTableName = delegationsV3Table
+
+  /**
+   * @param {import('../types/database').Database<AccessApiD1TablesV3>} db
+   * @param {R2Bucket} dags
+   * @param {keyof AccessApiD1TablesV3} delegationsTableName
+   */
+  // eslint-disable-next-line no-useless-constructor
+  constructor(db, dags, delegationsTableName = delegationsV3Table) {
+    this.#db = db
+    this.#delegationsTableName = delegationsTableName
+    this.#dags = dags
+  }
+
+  /**
+   * store items
+   *
+   * @param  {Array<Ucanto.Delegation>} delegations
+   * @returns {Promise<void>}
+   */
+  async putMany(...delegations) {
+    if (delegations.length === 0) {
+      return
+    }
+    await writeDelegations(this.#dags, delegations)
```

Comment:
We might succeed here, but then fail on the insert, in which case we will be storing more than is necessary in the R2Bucket, but we should 

If d1 had access to sql transactions, we might be able to use that here to do the write to r1 after trying the sql insert inside a transaction, but we don't so this is I think the best we can do without a much more involved [2pc](https://en.wikipedia.org/wiki/Two-phase_commit_protocol)